### PR TITLE
Kernel refactoring

### PR DESCRIPTION
**Pull request by thomaspinder**
Refactor the kernel module of gpjax to be compatible with PyTree.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

File: gpjax/parameters/bijectors.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    forward: Callable = static_field()
+    inverse: Callable = static_field()
+
+
+Identity = Bijector(forward=lambda x: x, inverse=lambda x: x)
+
+Softplus = Bijector(
+    forward=lambda x: jnp.log(1.0 + jnp.exp(x)),
+    inverse=lambda x: jnp.log(jnp.exp(x) - 1.0),
```
Comment (st--) [inline gpjax/parameters/bijectors.py:33]: These are not very numerically stable (and the forward will easily overflow). Better would be
```suggestion
    forward=jax.nn.softplus,
    inverse=tfp.math.softplus_inverse,
```
or, would it be an option to simply use tfp's bijector implementations directly?

File: gpjax/parameters/module.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+        Returns:
+            Module: tranformed to the unconstrained space.
+        """
+
+        def _apply_unconstrain(meta_leaf):
+            meta, leaf = meta_leaf
+            return meta.get("bijector", Identity).inverse(leaf)
+
+        return meta_map(_apply_unconstrain, self)
```
Comment (st--) [inline gpjax/parameters/module.py:128]: I find this confusing. What's the "base representation"? This makes me think that the bijector is effectively applied twice, i.e. `bijector.forward(bijector.forward(m.unconstrain())) == m.constrain()`, where I would expect the difference between "constrained" and "unconstrained" to be a _single_ application of the bijector :thinking:

File: gpjax/parameters/bijectors.py (hunk 1 above)
Comment (Daniel-Dodd) [reply gpjax/parameters/bijectors.py:33]: Nice, thanks @st-- 

Yeah, I favour using TFP. We will be dropping distrax in favour of this too.

(This is just legacy code from Mytree that aimed to be backend dependancy free).

**Comment by thomaspinder** [comment]
Subsumed by PRs into `v0.6`.

**Pull request closed without merging**